search-impl/src/test/java/com/example/core/TopicStub.java

```diff
@@ -49,13 +54,21 @@ public TopicStub(Materializer materializer) {
                         .via(flow)
                         .toMat(Sink.ignore(), Keep.both())
                         .run(materializer);
-                sourceActor = pair.first();
+                sourceActor.complete(pair.first());
                 return pair.second();
             }
         };
     }
 
-    public Supplier<ActorRef> actorSupplier() {
-        return () -> sourceActor;
+    /**
+     * @return a {@link ProducerStub} that clients (test code) can use to send messages through this {@link TopicStub}.
+     */
+    public ProducerStub<T> producerStub() {
+        return (T message) ->
+                sourceActor.thenApply(topicActor -> {
```

Comment:
I think there's a race condition here.  Let's say you have the following code:

```java
ProducerStub<String> stub = topicStub.producerStub();

stub.send("1");
stub.send("2");
```

In what order will those messages arrive?  I don't think it's defined.  They are sent by successive callbacks added to the `sourceActor` future.  I don't think there's any guarantees on ordering of callbacks attached to a `CompletableFuture`, in particular, let's say the first one attaches before the future is redeemed, then the future is redeemed, and at the same time the second attaches.  It may be that it executes before the first one.

I think the best way to implement this and guarantee safety will be to create a single actor that backs the `TopicStub` stub, which buffers messages before the stream is subscribed to, and once it's subscribed, then drain the queue and send messages.  You could also support multiple calls to subscribe, having multiple actors to send to, though of course only the first actor would be guaranteed to get all the buffered messages, subsequent subscriptions will only get messages published after they've subscribed.